src/runner.rs

```
@@ -6,511 +6,183 @@
 #[derive(Default)]
 struct Handlers<'a> {
-    handlers: Vec<&'a mut events::EventHandler>,
+    handlers: Vec<&'a mut EventHandler>,
 }
 
 impl<'a> Handlers<'a> {
-    fn broadcast(&mut self, event: &events::Event) {
+    fn broadcast(&mut self, event: &Event) {
         for h in &mut self.handlers {
             h.trigger(event)
         }
     }
 }
 
-pub struct Runner<'a> {
-    describe: Context<'a>,
-    report: Option<RunnerResult>,
+pub struct Runner<'a, T>
+    where T: 'a
+{
+    suite: Option<Suite<'a, T>>,
+    environments: Vec<T>,
+    report: ContextReport,
     handlers: Handlers<'a>,
 }
 
-impl<'a> Runner<'a> {
-    pub fn new(context: Context<'a>) -> Runner<'a> {
+impl<'a, T> Runner<'a, T> {
+    pub fn new(suite: Suite<'a, T>, environment: T) -> Runner<'a, T> {
         Runner {
-            describe: context,
+            suite: Some(suite),
+            environments: vec![environment],
+            report: ContextReport::default(),
             handlers: Handlers::default(),
-            report: None,
         }
     }
 }
 
-#[derive(Debug, PartialEq, Eq, Clone, Copy, Default)]
-pub struct TestReport {
-    pub total_tests: u32,
-    pub success_count: u32,
-    pub error_count: u32,
-}
-
-impl<'a> Runner<'a> {
-
-    #[cfg_attr(feature="clippy", allow(redundant_closure))]
-    fn run_test(mut test_function: Box<TestFunction>)
-                -> ExampleResult {
-
-        use std::panic::{catch_unwind, AssertUnwindSafe};
-        use example_result;
-
-        let res = catch_unwind(AssertUnwindSafe(|| test_function()));
-
-        res.unwrap_or_else(|_| example_result::FAILED_RES)
+impl<'a, T> Runner<'a, T>
+    where T: 'a + Clone + ::std::fmt::Debug
+{
+    pub fn run(mut self) -> ContextReport {
+        let suite = mem::replace(&mut self.suite, None).expect("Expected context");
+        panic::set_hook(Box::new(|_panic_info| {
+            // silently swallows panics
```

You mean the `// silently swallows panics`? Oh that just prevents the panic from getting printed to the console.

It's [this](https://github.com/mackwic/rspec/pull/30/files/f8be924fbe008473ffd29ec63ea61299043fe360#diff-e0934bb6f70a4f8c3baf22a773c14ec7R414) that does the actual fail catching.